src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs

```diff
@@ -3065,6 +3066,9 @@ private void SubstituteFakeTokens(Blob blob, ImmutableArray<byte> methodBodyIL,
                                         case LiteralGreatestMethodDefinitionToken:
                                             token = GreatestMethodDefIndex;
                                             break;
+                                        case SourceDocumentIndex:
+                                            token = _dynamicAnalysisDataWriterOpt.GetOrAddDocument(((CommonPEModuleBuilder)module).GetSourceDocumentFromFakeToken((uint)(pseudoToken & 0x00ffffff)));
```

Comment:
Rather than having `MetadataWriter` depend on `CommonPEModuleBuilder`, should `IModule` have an `IEnumerable<Cci.DebugSourceDocument> SourceDocumentsInIL(out int count)` method, similar to `ReferencesInIL()`?